File: exercise/ts/week05-complexity/exercise2/src/commandProcessor.ts
import {Command} from "./command";

export class CommandProcessor {
    private commandMap: Map<string, Command>;

    constructor() {
        this.commandMap = new Map<string, Command>();
        this.commandMap.set("greet", { execute: () => console.log("Hello, World!") });
        this.commandMap.set("exit", { execute: () => console.log("Exiting application...") });
    }

    processCommand(command: string): void {
        if (this.commandMap.has(command)) {
            this.commandMap.get(command)!.execute();
        } else {
            console.log("Unknown command");
        }
    }
}

const cp = new CommandProcessor();
// TODO: Should be able to pass my name to the command to say Hello to me!
cp.processCommand("greet");  // Outputs: Hello, World!
cp.processCommand("exit");   // Outputs: Exiting application...
// TODO: Should display all commands available
cp.processCommand("help");   // Outputs: Unknown command

File: exercise/c#/week05-complexity/CommandProcessor/CommandProcessor.cs
namespace CommandProcessor
{
    public class CommandProcessor
    {
        private readonly Dictionary<string, ICommand> _commandMap = new();

        public CommandProcessor()
        {
            _commandMap["greet"] = new ActionCommand(() => Console.WriteLine("Hello, World!"));
            _commandMap["exit"] = new ActionCommand(() => Console.WriteLine("Exiting application..."));
        }

        public void ProcessCommand(string command)
        {
            if (_commandMap.ContainsKey(command))
            {
                _commandMap[command]?.Execute();
            }
            else
            {
                Console.WriteLine("Unknown command");
            }
        }
    }

    public class ActionCommand(Action action) : ICommand
    {
        public void Execute() => action();
    }
}